File: ProblemSet/44. 最小子数组/java/44. 最小子数组 分治法.java
import java.util.List;

public class Solution {
    private List<Integer> nums;
    public class Status {
        public int lSum, rSum, mSum, tSum;

        public Status(int lSum, int rSum, int mSum, int tSum) {
            this.lSum = lSum;
            this.rSum = rSum;
            this.mSum = mSum;
            this.tSum = tSum;
        }
    }
    /*
     * @param nums: a list of integers
     * @return: A integer indicate the sum of minimum subarray
     */
    public int minSubArray(List<Integer> nums) {
        this.nums = nums;
        return getSumInfo(0, nums.size() - 1).mSum;
    }

    private Status getSumInfo(int l, int r) {
        if (l == r) {
            return new Status(nums.get(l), nums.get(l), nums.get(l), nums.get(l));
        }
        int m = l + ((r - l) >> 1);
        Status lStatus = getSumInfo(l, m);
        Status rStatus = getSumInfo(m + 1, r);
        int lSum = Math.min(lStatus.lSum, lStatus.tSum + rStatus.lSum);
        int rSum = Math.min(rStatus.rSum, rStatus.tSum + lStatus.rSum);
        int mSum = Math.min(lStatus.rSum + rStatus.lSum, Math.min(lStatus.mSum, rStatus.mSum));
        int tSum = lStatus.tSum + rStatus.tSum;
        return new Status(lSum, rSum, mSum, tSum);
    }
}

File: ProblemSet/44. 最小子数组/python/44. 最小子数组 分治法.py
class Status:
    def __init__(self, lSum, rSum, mSum, tSum):
        self.lSum = lSum
        self.rSum = rSum
        self.mSum = mSum
        self.tSum = tSum

class Solution:
    """
    @param: nums: a list of integers
    @return: A integer indicate the sum of minimum subarray
    """
    def minSubArray(self, nums):
        self.nums = nums
        return self.getSumInfo(0, len(nums) - 1).mSum
    
    def getSumInfo(self, l, r):
        if l == r:
            return Status(self.nums[l], self.nums[l], self.nums[l], self.nums[l])
        
        m = l + ((r - l) >> 1)
        lStatus = self.getSumInfo(l, m)
        rStatus = self.getSumInfo(m + 1, r)
        lSum = min(lStatus.lSum, lStatus.tSum + rStatus.lSum)
        rSum = min(rStatus.rSum, rStatus.tSum + lStatus.rSum)
        mSum = min(lStatus.rSum + rStatus.lSum, min(lStatus.mSum, rStatus.mSum))
        tSum = lStatus.tSum + rStatus.tSum
        return Status(lSum, rSum, mSum, tSum)